diff --git a/src/ParallelLaunch.cpp b/src/ParallelLaunch.cpp
index 6acb575..d3624b6 100644
--- a/src/ParallelLaunch.cpp
+++ b/src/ParallelLaunch.cpp
@@ -193,7 +193,7 @@ void parallelWork(GlobalTestConfig* gtc){
 		 * TODO will fix it.
 		 */
 
-		alarm(gtc->interval+30);  // set an alarm for interval+30 seconds from now
+		//alarm(gtc->interval+30);  // set an alarm for interval+30 seconds from now
 	}
 
 	// atomic_thread_fence(std::memory_order_acq_rel);
diff --git a/src/main.cpp b/src/main.cpp
index 00846a9..5890448 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -80,6 +80,7 @@
 #include "KVTest.hpp"
 #include "YCSBTest.hpp"
 #include "GraphTest.hpp"
+#include "GraphOpTest.hpp"
 
 #include "MapVerify.hpp"
 #include "QueueChurnTest.hpp"
@@ -182,6 +183,22 @@ int main(int argc, char *argv[])
 	gtc.addTestOption(new MapSyncTest<string, string>(50, 0, 25, 25, 1000000, 500000), "MapSyncTest<string>:g50p0i25rm25:range=1000000:prefill=500000");
 	gtc.addTestOption(new QueueSyncTest(50,50,2000), "QueueSync:eq50dq50:prefill=2000");
 
+	// Custom workload
+
+	gtc.addRideableOption(new MontageGraphFactory<5000, 1, 50>(), "MontageGraph-Custom");
+
+	gtc.addTestOption(new QueueTest(1000,50), "MontageQueue:1000");
+	gtc.addTestOption(new QueueTest(10000,50), "MontageQueue:10000");
+	gtc.addTestOption(new QueueTest(100000,50), "MontageQueue:100000");
+	gtc.addTestOption(new MapTest<string,string>(30, 30, 30, 10, 1000000, 1000, 1000), "MontageHashTable:1000");
+	gtc.addTestOption(new MapTest<string,string>(30, 30, 30, 10, 1000000, 1000, 10000), "MontageHashTable:10000");
+	gtc.addTestOption(new MapTest<string,string>(30, 30, 30, 10, 1000000, 1000, 100000), "MontageHashTable:100000");
+	gtc.addTestOption(new GraphOpTest(30,25,25,10,10, 1000000, 1000), "MontageGraph-Custom:1000");
+	gtc.addTestOption(new GraphOpTest(30,25,25,10,10, 1000000, 10000), "MontageGraph-Custom:10000");
+	gtc.addTestOption(new GraphOpTest(30,25,25,10,10, 1000000, 100000), "MontageGraph-Custom:100000");
+
+	
+
 	
 	gtc.addTestOption(new MapChurnTest<uint64_t,uint64_t>(50, 0, 25, 25, 1000000, 500000), "MapChurnTest<uint64_t>:g50p0i25rm25:range=1000000:prefill=500000");
 	gtc.addTestOption(new MapVerify<string, string>(50, 0, 25, 25, 1000000, 10000), "MapVerify");
diff --git a/src/tests/GraphOpTest.hpp b/src/tests/GraphOpTest.hpp
new file mode 100644
index 0000000..58b02be
--- /dev/null
+++ b/src/tests/GraphOpTest.hpp
@@ -0,0 +1,226 @@
+#ifndef GRAPHOPTEST_HPP
+#define GRAPHOPTEST_HPP
+
+/*
+ * This is a test with fixed number of operations for mappings.
+ */
+
+#include "TestConfig.hpp"
+#include "RGraph.hpp"
+#include <iostream>
+#ifdef PRONTO
+#include <signal.h>
+#include "savitar.hpp"
+#include "thread.hpp"
+#include "nvm_manager.hpp"
+#include "snapshot.hpp"
+#include <execinfo.h>
+#endif
+//KEY_SIZE and VAL_SIZE are only for string kv
+class GraphOpTest : public Test{
+#ifdef PRONTO
+    // some necessary var and func for running pronto
+    static pthread_t snapshot_thread;
+    static pthread_mutex_t snapshot_lock;
+
+    static void *snapshot_worker(void *arg) {
+        Snapshot *snap = (Snapshot *)arg;
+        snap->create();
+        delete snap;
+        return NULL;
+    }
+
+    static void signal_handler(int sig, siginfo_t *si, void *unused) {
+        assert(sig == SIGSEGV || sig == SIGUSR1);
+        if (sig == SIGSEGV) {
+            void *addr = si->si_addr;
+            if (!Snapshot::anyActiveSnapshot()) {
+                void *array[10];
+                size_t size;
+
+                size = backtrace(array, 10);
+                fprintf(stderr, "Segmentation fault!\n");
+                backtrace_symbols_fd(array, size, STDERR_FILENO);
+                exit(1);
+            }
+            Snapshot::getInstance()->pageFaultHandler(addr);
+        }
+        else { // SIGUSR1
+            pthread_mutex_lock(&snapshot_lock);
+            if (!Snapshot::anyActiveSnapshot()) {
+                Snapshot *snap = new Snapshot(PMEM_PATH);
+                pthread_create(&snapshot_thread, NULL, snapshot_worker, snap);
+            }
+            pthread_mutex_unlock(&snapshot_lock);
+        }
+    }
+#endif
+public:
+    int pge,pae,pre,pav,prv;
+    int getEdgeProb, addEdgeProb, remEdgeProb, addVerProb, remVerProb;
+    int totalProb;
+    int range;
+
+    RGraph *g;
+    size_t key_size = TESTS_KEY_SIZE;
+    size_t val_size = TESTS_VAL_SIZE;
+    std::string value_buffer; // for string kv only
+    uint64_t total_ops;
+    uint64_t* thd_ops;
+    GraphOpTest(int p_get_edge, int p_add_edge, int p_rem_edge, 
+                int p_add_vert, int p_rem_vert,
+      int range, int op = 10000000){
+
+        totalProb = p_get_edge;
+        pge = p_get_edge;
+        getEdgeProb = totalProb;
+        totalProb += p_add_edge;
+
+        pae = p_add_edge;
+        addEdgeProb = totalProb;
+        totalProb += p_rem_edge;
+        
+        pre = p_rem_edge;
+        remEdgeProb = totalProb;
+        totalProb += p_add_vert;
+
+        pav = p_add_vert;
+        addVerProb = totalProb;
+        totalProb += p_rem_vert;
+
+        prv = p_rem_vert;
+        remVerProb = totalProb;
+        
+        this->range = range;
+
+        total_ops = op;
+    }
+
+    void parInit(GlobalTestConfig* gtc, LocalTestConfig* ltc){
+        g->init_thread(gtc, ltc);
+    }
+    void init(GlobalTestConfig* gtc){
+#ifdef PRONTO
+        // init pronto things
+        Savitar_core_init();
+        NVManager::getInstance(); // recover persistent objects (blocking)
+
+        // Register signal handler for snapshots
+        pthread_mutex_init(&snapshot_lock, NULL);
+        struct sigaction sa;
+        memset(&sa, 0, sizeof(sa));
+        sa.sa_flags = SA_SIGINFO;
+        sigemptyset(&sa.sa_mask);
+        sa.sa_sigaction = signal_handler;
+        assert(sigaction(SIGSEGV, &sa, NULL) == 0);
+        assert(sigaction(SIGUSR1, &sa, NULL) == 0);
+#endif
+
+        getRideable(gtc);
+        
+        if(gtc->verbose){
+            printf("get_edge:%d add_edge:%d rem_edge:%d add_vertex:%d rm_vertex:%d\n",
+            pge,pae,pre,pav,prv);
+            printf("get_edge:%d add_edge:%d rem_edge:%d add_vertex:%d rm_vertex:%d sum:%d\n",
+            getEdgeProb,addEdgeProb,remEdgeProb,addVerProb,remVerProb, totalProb);
+        }
+        
+        // overrides for constructor arguments
+        if(gtc->checkEnv("range")){
+            range = atoi((gtc->getEnv("range")).c_str());
+        }
+
+        if(gtc->checkEnv("KeySize")){
+            key_size = atoi((gtc->getEnv("KeySize")).c_str());
+            assert(key_size<=TESTS_KEY_SIZE&&"KeySize dynamically passed in is greater than macro TESTS_KEY_SIZE!");
+        }
+        if(gtc->checkEnv("ValueSize")){
+            val_size = atoi((gtc->getEnv("ValueSize")).c_str());
+            assert(val_size<=TESTS_VAL_SIZE&&"ValueSize dynamically passed in is greater than macro TESTS_VAL_SIZE!");
+        }
+
+        value_buffer.reserve(val_size);
+        value_buffer.clear();
+        std::mt19937_64 gen_v(7);
+        for (size_t i = 0; i < val_size - 1; i++) {
+            value_buffer += (char)((i % 2 == 0 ? 'A' : 'a') + (gen_v() % 26));
+        }
+        value_buffer += '\0';
+
+
+        thd_ops = new uint64_t[gtc->task_num];
+        uint64_t new_ops = total_ops/gtc->task_num;
+        for(int i=0;i<gtc->task_num;i++){
+            thd_ops[i] = new_ops;
+        }
+        if(new_ops*gtc->task_num != total_ops) {
+            thd_ops[0] += (total_ops - new_ops*gtc->task_num);
+        }
+        /* set interval to inf so this won't be killed by timeout */
+        gtc->interval = numeric_limits<double>::max();
+    }
+
+    void getRideable(GlobalTestConfig* gtc){
+        Rideable* ptr = gtc->allocRideable();
+        g = dynamic_cast<RGraph*>(ptr);
+        if (!g) {
+             errexit("GraphOpTest must be run on RGraph<K,V> type object.");
+        }
+    }
+
+    int execute(GlobalTestConfig *gtc, LocalTestConfig *ltc) {
+        auto time_up = gtc->finish;
+        int ops = 0;
+        int tid = ltc->tid;
+        if (tid == 0) if(gtc->verbose) std::cout << "Starting test now..." << std::endl;
+        std::mt19937_64 gen_p(ltc->seed);
+        std::mt19937_64 gen_v(ltc->seed + 1);
+        std::uniform_int_distribution<> dist(0,totalProb);
+        std::uniform_int_distribution<> distv(0,5000);
+        auto now = std::chrono::high_resolution_clock::now();
+        
+        for (size_t i = 0; i < thd_ops[tid]; i++) {
+            int rng = dist(gen_p);
+            if (rng < getEdgeProb) {
+                g->has_edge(distv(gen_v), distv(gen_v));
+            } else if (rng < addEdgeProb) {
+                g->add_edge(distv(gen_v), distv(gen_v), -1);
+            } else if (rng < remEdgeProb) {
+                g->remove_edge(distv(gen_v), distv(gen_v));
+            } else if (rng < addVerProb) {
+                g->add_vertex(distv(gen_v));
+            } else {
+                g->remove_vertex(distv(gen_v));
+            }
+            ops++;
+        }
+        return ops;
+    }
+
+    void cleanup(GlobalTestConfig* gtc){
+#ifdef PRONTO
+        // Wait for active snapshots to complete
+        pthread_mutex_lock(&snapshot_lock);
+        if (Snapshot::anyActiveSnapshot()) {
+            pthread_join(snapshot_thread, NULL);
+        }
+        pthread_mutex_unlock(&snapshot_lock);
+
+        Savitar_core_finalize();
+        pthread_mutex_destroy(&snapshot_lock);
+#else
+        // Pronto handles deletion by its own
+        delete g;
+#endif
+    }
+
+};
+
+
+#ifdef PRONTO
+template<class K, class V>
+pthread_t GraphOpTest<K,V>::snapshot_thread;
+template<class K, class V>
+pthread_mutex_t GraphOpTest<K,V>::snapshot_lock;
+#endif
+#endif
\ No newline at end of file
diff --git a/src/tests/QueueTest.hpp b/src/tests/QueueTest.hpp
index 9444377..c2cdc5e 100644
--- a/src/tests/QueueTest.hpp
+++ b/src/tests/QueueTest.hpp
@@ -63,7 +63,7 @@ public:
     std::string trace_prefix;
     std::string thd_num;
     size_t val_size = TESTS_VAL_SIZE;
-    int prefill = 2000;
+    int prefill = 1000;
     uint64_t total_ops;
     uint64_t* thd_ops;
     unsigned int enq;
@@ -222,4 +222,4 @@ pthread_mutex_t QueueTest::snapshot_lock;
 //     cerr<<"QueueTest doesn't support string type!"<<std::endl;
 //     exit(1);
 // }
-#endif
\ No newline at end of file
+#endif
